pkg/rt/manager.go

```diff
@@ -68,32 +72,48 @@ func (m *Manager) Refresh() {
 	defer m.mtx.Unlock()
 
 	nextgen := map[string]interrupt{}
-	for _, id := range m.ids.ServiceIDs() {
-		if irq, ok := m.managed[id]; ok {
-			level.Debug(m.logger).Log("service_id", id, "subscriber", "maintain")
-			nextgen[id] = irq // move
-			delete(m.managed, id)
-		} else {
-			level.Info(m.logger).Log("service_id", id, "subscriber", "create")
-			nextgen[id] = m.spawn(id)
+	for _, product := range api.Products {
+		if m.productCache.HasAccess(product) {
+			for _, id := range m.ids.ServiceIDs() {
+				key := id + sep + product
```

Comment:
I'm not sure this is worth it, but you can define a struct with these key components instead of shoving them into a string:

```
type ProductKey struct {
    ServiceID string
    Product string
}
```

Instances of this struct can be directly used as keys to the `managed` and `nextgen` maps while also giving you direct access to the components without splitting strings. This lets us get rid of the string concatenation and `parseKey()`.

Reply:
Yeah, this is a really good idea. I wasn't happy with the separator and splitting the strings. I clean this up next week when I get back.